### PR TITLE
Feature jms with ActiveMQ

### PR DESCRIPTION
**Pull request by dilbertside**
Hi 
I contribute JMS (ActiveMQ implementation) to logstash plugin.
It has been working very nice for 1 month already in production, no problems.

Beside JMS, I added few line in BuildData.
I did that because one of the main use of this plugin was to feed another system which need a more extensive set of environment variables than the very limited ones provided by getBuildVariables.

Our jenkins job produce lots of dynamic EnvVars which are necessary to forward to other systems.
It maybe possible to make it optional, but I' m not familiar with Jenkins Jelly.

Unit Tests run fine.

Thanks to make available to the community this plugin.
Regards,

### CODE REVIEW

**Comment by dilbertside** [comment]
I don' t understand Travis CI failure

on my machine,I get with 
```
mvn clean package -Pjenkins
```
Results :
Tests run: 62, Failures: 0, Errors: 0, Skipped: 0

Is it with a different jvm?

File: src/main/java/jenkins/plugins/logstash/persistence/ActiveMqDao.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+      producer.setDeliveryMode(DeliveryMode.NON_PERSISTENT);
+      
+      // Create a messages
+      TextMessage message = session.createTextMessage(data);
+
+      // Tell the producer to send the message
+      producer.send(message);
+
+      // Clean up
+      session.close();
```
Comment (olamy) [inline src/main/java/jenkins/plugins/logstash/persistence/ActiveMqDao.java]: maybe having those close statement in a finally block?

Comment (dilbertside) [inline src/main/java/jenkins/plugins/logstash/persistence/ActiveMqDao.java]: Possible but that may lead to ugly code and redundancy

Methods Signature are like following
Connection
void close() throws JMSException;

Session
void close() throws JMSException;

Do you want that?

  public long push(String data, PrintStream logger) {
    Connection connection = null;
    Session session = null;
    try {
      // Create a Connection
      connection = connectionFactory.createConnection();
      connection.start();
      // Create a Session
      session = connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
     
      // Create the destination (Topic or Queue)
      Destination destination = session.createQueue(key);
     
      // Create a MessageProducer from the Session to the Topic or Queue
      MessageProducer producer = session.createProducer(destination);
      producer.setDeliveryMode(DeliveryMode.NON_PERSISTENT);
     
      // Create a messages
      TextMessage message = session.createTextMessage(data);

      // Tell the producer to send the message
      producer.send(message);

    } catch (JMSException e) {
      logger.println(ExceptionUtils.getStackTrace(e));
    } finally{
      try {
        session.close();
        connection.close();
        // Clean up
        return 1;
      } catch (JMSException e) {
        logger.println(ExceptionUtils.getStackTrace(e));
      }
    }
    return -1;
  }

**Comment by dilbertside** [comment]
I tried twice to read the page
http://jenkins-ci.org/pull-request-greeting (Fedora Firefox 37),
Waiting minutes, and I only got a blank page, doesn't work from Asia?

**Comment by R-Gerard** [comment]
The Travis CI build failed on JDK 1.6. It looks like the ActiveMQ client jar is compiled on JDK 1.7: 
https://travis-ci.org/jenkinsci/logstash-plugin/jobs/57162166#L171

It's a good idea to keep backwards compatibility with Java 6 because that's all that's required to run the Jenkins CI server.

File: pom.xml [hunk 2]
```diff
@@ -120,6 +126,10 @@
           </excludes>
         </configuration>
       </plugin>
+      <plugin>
+          <groupId>org.jenkins-ci.tools</groupId>
+          <artifactId>maven-hpi-plugin</artifactId>
+        </plugin>
```
Comment (R-Gerard) [inline pom.xml]: What is this change for?

File: src/main/java/jenkins/plugins/logstash/persistence/ActiveMqDao.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+import javax.jms.TextMessage;
+
+import org.apache.activemq.ActiveMQConnectionFactory;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.lang.exception.ExceptionUtils;
+
+/**
+ * ActiveMq Data Access Object.
+ *
+ * @author Rusty Gerard
```
Comment (R-Gerard) [inline src/main/java/jenkins/plugins/logstash/persistence/ActiveMqDao.java]: You don't need to include me as an author.

File: src/main/java/jenkins/plugins/logstash/persistence/ActiveMqDao.java (hunk 1 above)
Comment (R-Gerard) [inline src/main/java/jenkins/plugins/logstash/persistence/ActiveMqDao.java]: I agree with @olamy. It's probably best to have the cleanup logic in the finally block.

File: src/main/java/jenkins/plugins/logstash/persistence/LogstashIndexerDao.java [hunk 4]
```diff
@@ -39,7 +39,8 @@
   static enum IndexerType {
     REDIS,
     RABBIT_MQ,
-    ELASTICSEARCH
+    ELASTICSEARCH,
+    JMS
```
Comment (R-Gerard) [inline src/main/java/jenkins/plugins/logstash/persistence/LogstashIndexerDao.java]: Maybe call it "ACTIVE_MQ" instead? These values are used to populate the dropdown list in the global configuration.

File: src/main/resources/jenkins/plugins/logstash/LogstashInstallation/help-key.html [hunk 5]
```diff
@@ -1,5 +1,6 @@
 <div>
   <p>REDIS: The name of a Redis list or channel.<br/>
   RABBIT_MQ: The name of a RabbitMq queue.<br/>
-  ELASTICSEARCH: The name and type path. Example: "/indexName/type"</p>
+  ELASTICSEARCH: The name and type path. Example: "/indexName/type"<br/>
+  JMS: The name of a JMS queue.</p>
```
Comment (R-Gerard) [inline src/main/resources/jenkins/plugins/logstash/LogstashInstallation/help-key.html]: Again, perhaps "ACTIVE_MQ" instead of "JMS" here.

File: src/main/java/jenkins/plugins/logstash/persistence/BuildData.java [hunk 6]
```diff
@@ -139,6 +141,18 @@ public BuildData(AbstractBuild<?, ?> build, Date currentTime) {
...
+    List<Environment> environmentList = build.getEnvironments();
+    if (environmentList != null) {
+      for (Environment e : environmentList) {
+        if (e != null) {
+          e.buildEnvVars(env);
+          if(!env.isEmpty())
+            buildVariables.putAll(env);
+        }
+      }
+    }
```
Comment (R-Gerard) [inline src/main/java/jenkins/plugins/logstash/persistence/BuildData.java]: Can you submit a separate pull request for this change? That way this feature can be merged in regardless of the ActiveMQ support.
